### PR TITLE
DEV-3975 | Contributor portal does not show imported legacy subscriptions

### PR DESCRIPTION
**Pull request by benjaminEwhite**
Please fill out each section even if it's just with "N/A"

#### What's this PR do?

The primary objective of this PR is to solve for a bug that resulted in contributors not being able to see some expected recurring contributions, which blocks them from updating payment method or cancelling. 

Contributions were being filtered out of that response by the API because they had a null value set for payment type property (see explanation in JIRA ticket).

This PR solves for this by implementing a path to assign a non-null value for payment type for PIs associated with legacy imported recurring contributions. 

The underlying problem was that  our code had faulty implicit assumptions about what payment intents resulting from legacy import look like. We expected them to have a payment method object (in expanded form) but there are no guarantees that PIs retrieved from Stripe will meet this expectation.

In particular, all NRE-generated payment intents (that are not tied to flagged recurring contributions) should meet this expectation because of how we confirm those payments on frontend using payment element. However, it appears that the imported payment intents we are getting were created in such a way where this expectation is not met.

In that case, there is another place we can derive the payment method from, and that allows us to assign a value for `payment_type` which is derived from payment method.

Solving for the above introduced a new bug around updating pis in cache — for these uninvoiced subs (treated as PIs), the operations to update cache were initially unsuccessful. Everything looked right in Stripe, but our cache had stale data. I went ahead and solved for that.

In addition to solving for the underlying bug, I took the opportunity to retire some unittest implemented tests.

#### Why are we doing this? How does it help us?

Solves for critical bug that stops some contributors from being able to update or cancel their recurring contributions.

#### Are there detailed, specific, step-by-step testing instructions in the associated ticket?

Yes.

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?

No

#### Have automated unit tests been added? If not, why?

Yes

#### How should this change be communicated to end users?

We might want to communicate to any orgs that have had legacy subscriptions imported.

#### Are there any smells or added technical debt to note?

I discovered a bug with the .refunded property on StripePaymentIntent, and created a [follow up issue for it.](https://news-revenue-hub.atlassian.net/browse/DEV-3987).

#### Has this been documented? If so, where?

No

#### What are the relevant tickets? Add a link to any relevant ones.

- [DEV-3975](https://news-revenue-hub.atlassian.net/browse/DEV-3975)
- [DEV-3987](https://news-revenue-hub.atlassian.net/browse/DEV-3987)

#### Do any changes need to be made before deployment to production (adding environment variables, for example)? If so, open a ticket and link it to the ticket for this PR and list it here:

No

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

[DEV-3987](https://news-revenue-hub.atlassian.net/browse/DEV-3987) - to fix bug with refunded contributions appearing as paid.

[DEV-3975]: https://news-revenue-hub.atlassian.net/browse/DEV-3975?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/contributions/stripe_contributions_provider.py [hunk 1]
```diff
@@ -342,6 +380,22 @@ def serialize(self, payment_intents: list[stripe.PaymentIntent]) -> dict[str, di
                 logger.warning("Unable to process Contribution [%s]", pi.id, exc_info=ex)
         return data
 
+    def convert_uninvoiced_subs_into_contributions(
```
Comment (benjaminEwhite) [inline apps/contributions/stripe_contributions_provider.py:383]: This is drying up some logic that lived inline but is now valuable in two places.

File: apps/contributions/stripe_contributions_provider.py [hunk 2]
```diff
@@ -53,17 +54,48 @@ class StripePaymentIntent:
     CANCELABLE_STATUSES = ["trialing", "active", "past_due"]
     MODIFIABLE_STATUSES = ["incomplete", "trialing", "active", "past_due"]
 
+    DUMMY_CARD = AttrDict(**{"brand": None, "last4": None, "exp_month": None, "exp_year": None})
+
     def __init__(self, payment_intent):
         self.payment_intent = payment_intent
 
+    @property
+    def payment_method(self) -> StripeObject | None:
```
Comment (benjaminEwhite) [inline apps/contributions/stripe_contributions_provider.py:63]: Moved this into its own property. Some of this was inlined in other places.

File: apps/contributions/stripe_contributions_provider.py [hunk 3]
```diff
@@ -53,17 +54,48 @@ class StripePaymentIntent:
...
+            most_recent = max(self.payment_intent.charges.data, key=lambda x: x.created)
+            pm = most_recent.payment_method_details
+
+        return pm
+
     @property
     def invoice_line_item(self):
+        default = AttrDict({})
         if not self.payment_intent.invoice:
-            return [{}]
```
Comment (benjaminEwhite) [inline apps/contributions/stripe_contributions_provider.py:62]: Returning `[{}]` was a mistake in the original. This should only ever return a single item, not an array of items.

File: apps/contributions/views.py [hunk 4]
```diff
@@ -520,25 +520,47 @@ def update_subscription_and_pi_in_cache(
...
+            pi_cache_provider.upsert_uninvoiced_subscriptions(
+                pi_cache_provider.convert_uninvoiced_subs_into_contributions([subscription])
+            )
+
+    @staticmethod
+    def update_uninvoiced_subscription_in_cache(email: str, stripe_account_id: str, subscription: stripe.Subscription):
+        """Used to update respective caches after a subscription is updated or canceled."""
+        logger.info("Updating caches for subscription %s", subscription.id)
+        sub_cache_provider = SubscriptionsCacheProvider(email, stripe_account_id)
         sub_cache_provider.upsert([subscription])
```
Comment (njharman) [inline apps/contributions/views.py:547]: Not covered by unit tests

**Review by njharman — CHANGES_REQUESTED** [review]
> I took the opportunity to retire some unittest implemented tests.
+1

Changes, questions/suggestions, some missing coverage

File: apps/contributions/stripe_contributions_provider.py [hunk 5]
```diff
@@ -158,24 +190,29 @@ def credit_card_expiration_date(self):
...
         or acount_refunded will be > 0 (will be useful in case of partial refund and we still want to set
         the status as refunded)
         https://stripe.com/docs/api/charges/object#charge_object-refunded
         https://stripe.com/docs/api/charges/object#charge_object-amount_refunded
         """
-        return self.payment_intent.get("refunded", False) or self.payment_intent.get("amount_refunded", 0) > 0
+        if "refunded" in self.payment_intent:
+            return self.payment_intent.refunded
+        if "amount_refunded" in self.payment_intent:
+            return self.payment_intent.amount_refunded > 0
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/contributions/stripe_contributions_provider.py:213]: This entire method is broken. See the todo at line 202 above. We'll fix in a separate ticket.

File: apps/contributions/stripe_contributions_provider.py [hunk 6]
```diff
@@ -53,17 +54,48 @@ class StripePaymentIntent:
...
     MODIFIABLE_STATUSES = ["incomplete", "trialing", "active", "past_due"]
 
+    DUMMY_CARD = AttrDict(**{"brand": None, "last4": None, "exp_month": None, "exp_year": None})
+
     def __init__(self, payment_intent):
         self.payment_intent = payment_intent
 
+    @property
+    def payment_method(self) -> StripeObject | None:
+        pm = None
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/contributions/stripe_contributions_provider.py:64]: e1ba75df

File: apps/contributions/views.py [hunk 7]
```diff
@@ -520,25 +520,47 @@ def update_subscription_and_pi_in_cache(
...
+        pi_id = subscription.latest_invoice.payment_intent if subscription.latest_invoice else None
+        pi = (
+            stripe.PaymentIntent.retrieve(
+                pi_id,
+                stripe_account=stripe_account_id,
+                expand=["payment_method", "invoice.subscription.default_payment_method"],
+            )
+            if pi_id
+            else None
         )
```
[the remark this answers is not in the document]
Comment (njharman) [reply apps/contributions/views.py:563]: good

File: apps/contributions/stripe_contributions_provider.py [hunk 8]
```diff
@@ -94,7 +126,7 @@ def interval(self):
     def revenue_program(self):
         metadata = self.payment_intent.get("metadata") or self.invoice_line_item.get("metadata") or {}
         if not metadata or "revenue_program_slug" not in metadata:
-            raise InvalidMetadataError(f"Metadata is invalid for payment_intent : {self.id}")
+            raise InvalidMetadataError(f"Metadata is invalid for payment_intent : {self.id}, %s", metadata)
```
[the remark this answers is not in the document]
Comment (njharman) [reply apps/contributions/stripe_contributions_provider.py]: Darn, I was hoping there was some cool, obscure reason why.

**Review by njharman — APPROVED** [review]

**Review by x110dc — APPROVED** [review]

**Pull request merged**